oap-server/server-configuration/configuration-etcd/src/main/java/org/apache/skywalking/oap/server/configuration/etcd/EtcdConfigWatcherRegister.java

```diff
@@ -0,0 +1,104 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.configuration.etcd;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Set;
+import mousio.etcd4j.EtcdClient;
+import mousio.etcd4j.promises.EtcdResponsePromise;
+import mousio.etcd4j.responses.EtcdErrorCode;
+import mousio.etcd4j.responses.EtcdException;
+import mousio.etcd4j.responses.EtcdKeysResponse;
+import org.apache.skywalking.oap.server.configuration.api.ConfigTable;
+import org.apache.skywalking.oap.server.configuration.api.ConfigWatcherRegister;
+import org.apache.skywalking.oap.server.library.module.ModuleStartException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Alan Lau
+ */
+public class EtcdConfigWatcherRegister extends ConfigWatcherRegister {
+
+    private final static Logger logger = LoggerFactory.getLogger(EtcdConfigWatcherRegister.class);
+
+    private EtcdServerSettings settings;
+
+    private EtcdClient client;
+
+    @Override public ConfigTable readConfig(Set<String> keys) {
+
+        if (client == null) {
+            try {
+                client = new EtcdClient(EtcdUtils.parse(settings).toArray(new URI[] {}));
+            } catch (ModuleStartException e) {
+                logger.error(e.getMessage(), e);
+            }
+
+            EtcdResponsePromise<EtcdKeysResponse> promise;
+            try {
+                promise = client.get(settings.getClusterName()).waitForChange().send();
+                promise.addListener(responsePromise -> {
+                    onDataValueChanged();
+                });
+
+            } catch (IOException e) {
+                logger.error(e.getMessage(), e);
+            }
+        }
+
+        final ConfigTable table = new ConfigTable();
+        keys.forEach(registryKey -> {
+            String key = new StringBuilder("/").append(settings.getGroup()).append("/").append(registryKey).toString();
+            try {
+                EtcdResponsePromise<EtcdKeysResponse> promise = client.get(key).send();
+                EtcdKeysResponse response = promise.get();
+                table.add(new ConfigTable.ConfigItem(getRealKey(key, settings.getGroup()), response.getNode().getValue()));
+            } catch (EtcdException e) {
+                if (e.getErrorCode() == EtcdErrorCode.KeyNotFound) {
+                    table.add(new ConfigTable.ConfigItem(getRealKey(key, settings.getGroup()), null));
+                } else {
+                    logger.error(e.getMessage(), e);
+                }
+            } catch (Exception e1) {
+                logger.error(e1.getMessage(), e1);
+            }
+        });
+
+        return table;
+    }
+
+    public EtcdConfigWatcherRegister(EtcdServerSettings settings) {
+        this.settings = settings;
+    }
+
+    private void onDataValueChanged() {
```

Comment:
What is this empty method for?

Reply:
Actually, it is unused now. Maybe use to monitor the value changed some time for some thing. But i didn't impl that at this time.

Reply:
This is a private method. So you write it by yourself. No one asks for this.

Reply:
removed.